loadgen/generic_executor.go

```diff
@@ -106,24 +127,30 @@ func (g *genericRun) Run(ctx context.Context) error {
 		go func() {
 			startTime := time.Now()
 			err := g.executor.Execute(ctx, run)
-			// Only log/wrap/send to channel if context is not done
-			if ctx.Err() == nil {
-				if err != nil {
-					err = fmt.Errorf("iteration %v failed: %w", run.Iteration, err)
-					g.logger.Error(err)
-				}
-				select {
-				case <-ctx.Done():
-				case doneCh <- err:
-					// Record/log here, not if it was cut short by context complete
-					g.executeTimer.Record(time.Since(startTime))
-				}
+			if err != nil {
+				err = fmt.Errorf("iteration %v failed: %w", run.Iteration, err)
+				g.logger.Error(err)
+			}
+			select {
+			case <-ctx.Done():
+			case doneCh <- err:
+				// Record/log here, not if it was cut short by context complete
+				g.executeTimer.Record(time.Since(startTime))
 			}
 		}()
 	}
-	// Wait for all to be done or an error to occur
-	for runErr == nil && ctx.Err() == nil && currentlyRunning > 0 {
-		waitOne()
+	g.logger.Debugf("No longer starting new iterations")
+	if currentlyRunning == 0 {
+		return fmt.Errorf("no runs were started. Iterations should be >0 or duration should be >0")
+	}
+	// Wait for all to be done or an error to occur. We will wait past the overall duration for
+	// executions to complete. It is expected that whatever is running omes may choose to enforce
+	// a hard timeout if waiting for started executions to complete exceeds a certain threshold.
+	for runErr == nil && currentlyRunning > 0 && ctx.Err() == nil {
+		waitOne(false)
+	}
+	if ctx.Err() != nil {
+		return fmt.Errorf("timed out while waiting for runs to complete: %w", ctx.Err())
```

Comment:
nit: move this termination into the loop? (Technically there's a tiny race as it is)